sentinel-core/src/main/java/com/alibaba/csp/sentinel/util/TimeUtil.java

```diff
@@ -15,38 +15,199 @@
  */
 package com.alibaba.csp.sentinel.util;
 
+import java.util.List;
 import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLong;
+
+import com.alibaba.csp.sentinel.log.RecordLog;
+import com.alibaba.csp.sentinel.slots.statistic.base.LeapArray;
+import com.alibaba.csp.sentinel.slots.statistic.base.LongAdder;
+import com.alibaba.csp.sentinel.slots.statistic.base.WindowWrap;
+import com.alibaba.csp.sentinel.util.function.Tuple2;
 
 /**
  * Provides millisecond-level time of OS.
+ * <p>
+ * Here we should see that not all the time TimeUtil should 
+ * keep looping 1_000 times every second (Actually about 800/s due to some losses).
+ * <pre>
+ * * In idle conditions it just acts as System.currentTimeMillis();
+ * * In busy conditions (significantly more than 1_000/s) it keeps loop to reduce costs.
+ * </pre>
+ * For detail design and proposals please goto 
+ * <a href="https://github.com/alibaba/Sentinel/issues/1702#issuecomment-692151160">https://github.com/alibaba/Sentinel/issues/1702</a>
  *
  * @author qinan.qn
+ * @author jason
  */
-public final class TimeUtil {
+public final class TimeUtil implements Runnable {
+    private static final long CHECK_INTERVAL = 3000;
+    
+    public static enum STATE {
+        IDLE, PREPARE, RUNNING;
+    }
+    private static class Statistic {
+        private LongAdder writes = new LongAdder();
+        private LongAdder reads = new LongAdder();
+        public LongAdder getWrites() {
+            return writes;
+        }
+        public LongAdder getReads() {
+            return reads;
+        }
+    }
+    private static TimeUtil INSTANCE;
 
-    private static volatile long currentTimeMillis;
+    private volatile long currentTimeMillis;
+    private AtomicLong lastCheck = new AtomicLong();
+    private LeapArray<Statistic> statistics;
+    private STATE state = STATE.IDLE;
```

Comment:
I think state should be a volatile variable.

Reply:
Hi, liqiangz

Here to use modifiers without a `volatile` is because following considerations:

* Low frequency of switching is assumed.
* All writings to it only occurred in same thread (which maybe we should consider to remove the unnecessary CAS operation
* Use L1/L2 cache to get better performance

So should it still be volatile? Is there any other branch I maybe lack of consideration?

Reply:
Thanks for your reply, but I still have a confusion.  When state in the daemon thread switch to IDLE or PREPARE, currentTimeMillis will not be updated. But when calling currentTime function  in another thread, state may still be RUNNING.  The `getTime()`  result will always be the old currentTimeMillis.

Reply:
Hi, liqiangz

Both IDLE and PREPARE are treated as IDLE in `getTime()` while daemon will only switch PREPARE to RUNNING when it actually started to update `currentTimeMillis`.

I will remove the CAS operation which is a kind of legacy after several iterations of design. And I also will consider whether it's necessary to make it volatile in java for not strictly reordering forbidden scenario. As we know in rare condition codes like below in c++:

```c++
bool cond = true;
while (cond);
```

may be "optimized" to be 

```c++
if (cond) {
while(true){}
}
```

But I am not sure it will occur in JVM's byte codes. Because it more like a bug which it's common in CPP compilers. (aha)

What do you think about the visibility latency in practice?

Reply:
CAS operation has been removed and state is taken as volatile now.
In unit test concurrency has been raised more too.

Reply:
I think we should be very careful about variables shared by multiple threads. If you want to do some concurrency optimization,  you must really understand what you do.

For programs that are not synchronized, if you want to judge the execution order of memory operations, it is almost impossible to get a correct conclusion . A simple way is to synchronize all variables shared by multiple threads.

In this scenario, the value of state modified by one thread may never be seen by another thread, because we don’t know what kind of optimization the compiler and cpu will perform. But because currentTimeMillis is volatile, There may be no problems in this program.

Reply:
Yeah totally agree with you. So CAS has been removed and volatile is marked on state while lastCheck is still a private and local variable.